assets/src/stories-editor/components/media-inserter/index.js

```diff
@@ -0,0 +1,194 @@
+			type: mediaType,
+		},
+	} );
+	let attachment;
+
+	// When an image is selected, run a callback.
+	fileFrame.on( 'select', () => {
+		attachment = fileFrame.state().get( 'selection' ).first().toJSON();
+		updateBlock( attachment );
+	} );
+
+	// Finally, open the modal
+	fileFrame.open();
+};
+
+const applyWithSelect = withSelect( ( select ) => {
+	const { getCurrentPage } = select( 'amp/story' );
+	const { canInsertBlockType, getBlockListSettings } = select( 'core/block-editor' );
+	const { isReordering } = select( 'amp/story' );
+
+	return {
+		isReordering: isReordering(),
+		canInsertBlockType: ( name ) => {
+			// canInsertBlockType() alone is not enough, see https://github.com/WordPress/gutenberg/issues/14515
+			const blockSettings = getBlockListSettings( getCurrentPage() );
+			return canInsertBlockType( name, getCurrentPage() ) && blockSettings && blockSettings.allowedBlocks.includes( name );
+		},
+		// As used in <HeaderToolbar> component
+		showInserter: select( 'core/edit-post' ).getEditorMode() === 'visual' && select( 'core/editor' ).getEditorSettings().richEditingEnabled,
+	};
+} );
+
+const applyWithDispatch = withDispatch( ( dispatch, props, { select } ) => {
+	const { getCurrentPage } = select( 'amp/story' );
+	const { getBlockOrder } = select( 'core/block-editor' );
+	const { insertBlock } = dispatch( 'core/block-editor' );
+
+	return {
+		insertBlock: ( name ) => {
+			const currentPage = getCurrentPage();
+			const index = getBlockOrder( currentPage ).length;
+
+			const insertedBlock = createBlock( name, {} );
+
+			insertBlock( insertedBlock, index, currentPage );
+		},
+		updateBlock: ( media ) => {
+			const clientId = getCurrentPage();
+			const { updateBlockAttributes, selectBlock } = dispatch( 'core/block-editor' );
+
+			if ( ! clientId ) {
+				return;
+			}
+
+			if ( ! media || ! media.url ) {
+				updateBlockAttributes( clientId,
+					{
+						mediaUrl: undefined,
+						mediaId: undefined,
+						mediaType: undefined,
+						mediaAlt: undefined,
+						poster: undefined,
+					}
+				);
+				return;
+			}
+
+			let mediaType;
+
+			// For media selections originated from a file upload.
+			if ( media.media_type ) {
+				if ( media.media_type === VIDEO_BACKGROUND_TYPE ) {
+					mediaType = VIDEO_BACKGROUND_TYPE;
+				} else {
+					mediaType = IMAGE_BACKGROUND_TYPE;
+				}
+			} else {
+				// For media selections originated from existing files in the media library.
+				if (
+					media.type !== IMAGE_BACKGROUND_TYPE &&
+					media.type !== VIDEO_BACKGROUND_TYPE
+				) {
+					return;
+				}
+
+				mediaType = media.type;
+			}
+
+			const mediaAlt = media.alt || media.title;
+			const mediaUrl = media.url;
+			const poster = VIDEO_BACKGROUND_TYPE === mediaType && media.image && media.image.src !== media.icon ? media.image.src : undefined;
+			updateBlockAttributes( clientId, {
+				mediaUrl,
+				mediaId: media.id,
+				mediaType,
+				mediaAlt,
+				poster,
+			} );
+			selectBlock( clientId );
```

Comment:
This seems to be largely matching the logic from `onSelectMedia` of the Page block -- could we create a helper function instead for the matching part and move it to `helpers` to reduce duplication?

Reply:
Created a new helper function called `processMedia`.